Mathlib/Topology/Algebra/RestrictedProduct.lean

```diff
@@ -0,0 +1,769 @@
+  rw [le_principal_iff] at hS
+  convert isOpen_set_pi (hS.inter_of_left {i | p i}) (fun i _ ↦ hAopen i) |>.preimage continuous_coe
+  ext f
+  refine ⟨fun H i hi ↦ H i hi.2, fun H i hiT ↦ ?_⟩
+  by_cases hiS : i ∈ S
+  · exact f.2 hiS
+  · exact H i ⟨hiS, hiT⟩
+
+include hAopen in
+theorem isOpen_forall_mem_of_principal {S : Set ι} (hS : cofinite ≤ 𝓟 S) :
+    IsOpen {f : Πʳ i, [R i, A i]_[𝓟 S] | ∀ i, f.1 i ∈ A i} := by
+  convert isOpen_forall_imp_mem_of_principal hAopen hS (p := fun _ ↦ True)
+  simp
+
+include hAopen in
+theorem isOpen_forall_imp_mem {p : ι → Prop} :
+    IsOpen {f : Πʳ i, [R i, A i] | ∀ i, p i → f.1 i ∈ A i} := by
+  simp_rw [topologicalSpace_eq_iSup cofinite, isOpen_iSup_iff, isOpen_coinduced]
+  exact fun S hS ↦ isOpen_forall_imp_mem_of_principal hAopen hS
+
+include hAopen in
+theorem isOpen_forall_mem :
+    IsOpen {f : Πʳ i, [R i, A i] | ∀ i, f.1 i ∈ A i} := by
+  simp_rw [topologicalSpace_eq_iSup cofinite, isOpen_iSup_iff, isOpen_coinduced]
+  exact fun S hS ↦ isOpen_forall_mem_of_principal hAopen hS
+
+include hAopen in
+theorem isOpenEmbedding_inclusion_principal {S : Set ι} (hS : cofinite ≤ 𝓟 S) :
+    IsOpenEmbedding (inclusion R A hS) where
+  toIsEmbedding := isEmbedding_inclusion_principal hS
+  isOpen_range := by
+    rw [range_inclusion]
+    exact isOpen_forall_imp_mem hAopen
+
+include hAopen in
+/-- `Π i, A i` is homeomorphic to an open subset of the restricted product. -/
+theorem isOpenEmbedding_structureMap :
+    IsOpenEmbedding (structureMap R A cofinite) where
+  toIsEmbedding := isEmbedding_structureMap
+  isOpen_range := by
+    rw [range_structureMap]
+    exact isOpen_forall_mem hAopen
+
+include hAopen in
+theorem nhds_eq_map_inclusion {S : Set ι} (hS : cofinite ≤ 𝓟 S)
+    (x : Πʳ i, [R i, A i]_[𝓟 S]) :
+    (𝓝 (inclusion R A hS x)) = map (inclusion R A hS) (𝓝 x) := by
+  rw [isOpenEmbedding_inclusion_principal hAopen hS |>.map_nhds_eq x]
+
+include hAopen in
+theorem nhds_eq_map_structureMap
+    (x : Π i, A i) :
+    (𝓝 (structureMap R A cofinite x)) = map (structureMap R A cofinite) (𝓝 x) := by
+  rw [isOpenEmbedding_structureMap hAopen |>.map_nhds_eq x]
+
+include hAopen in
+/-- If each `R i` is weakly locally compact, each `A i` is open, and all but finitely many `A i`s
+are also compact, then the restricted product `Πʳ i, [R i, A i]` is weakly locally compact. -/
+theorem weaklyLocallyCompactSpace_of_cofinite [∀ i, WeaklyLocallyCompactSpace (R i)]
+    (hAcompact : ∀ᶠ i in cofinite, IsCompact (A i)) :
+    WeaklyLocallyCompactSpace (Πʳ i, [R i, A i]) where
+  exists_compact_mem_nhds := fun x ↦ by
+    set S := {i | IsCompact (A i) ∧ x i ∈ A i}
+    have hS : cofinite ≤ 𝓟 S := le_principal_iff.mpr (hAcompact.and x.2)
+    have hSx : ∀ i ∈ S, x i ∈ A i := fun i hi ↦ hi.2
+    have hSA : ∀ i ∈ S, IsCompact (A i) := fun i hi ↦ hi.1
+    haveI := weaklyLocallyCompactSpace_of_principal hS hSA
+    rcases exists_inclusion_eq_of_eventually R A hS hSx with ⟨x', hxx'⟩
```

Comment:
I thought all the cool kids did `obtain` nowadays (on the basis that it is a word, and `rcases` isn't).

Reply:
I agree with the workflow, except that it works just fine with `obtain`